### PR TITLE
feat: Optimism builder & miner

### PR DESCRIPTION
**Pull request by Wodann**

### CODE REVIEW

**Comment by fvictorio** [comment]
This PR seems to have unrelated changes. Maybe you accidentally merged `main` into the `refactor/optimism-builder` branch?

**Comment by fvictorio** [comment]
I'm not sure. If that's too much work, it might not be worth it. But this PR as it is is really hard to review.

Perhaps you can list the commits or commit ranges that need reviewing?

In any case, all the changes in areas that I usually review seem to come from `main`, so I'll leave it up to @agostbiro to decide how he prefers to review this.

File: crates/edr_opt/tests/full_block.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+impl_full_block_tests! {
+    // This test often failes with an internal Alchemy error:
+    // > execution aborted (timeout = 10s)
+    //
+    // Potentially the block is too old?
+    // mainnet_pre_bedrock => OptimismChainSpec {
+    //     block_number: 98_235_064,
+    //     chain_id: 10,
+    //     url: get_alchemy_url().replace("eth-", "opt-"),
+    // },
```
Comment (agostbiro) [inline crates/edr_opt/tests/full_block.rs:16]: Should this be removed?

File: crates/edr_opt/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+[package]
+name = "edr_opt"
```
Comment (agostbiro) [inline crates/edr_opt/Cargo.toml:2]: It took me a moment to parse the name, I'd prefer `edr_optimism` if possible

File: crates/edr_evm/src/chain_spec.rs [hunk 3]
```diff
@@ -1,55 +1,140 @@
...
-impl<ChainSpecT> SyncChainSpec for ChainSpecT where
-    ChainSpecT: ChainSpec<Transaction: Send + Sync> + Send + Sync + 'static
-{
-}
+    /// Type representing an error that occurs when converting an RPC receipt.
+    type RpcReceiptConversionError: Debug + std::error::Error;
+
+    /// Type representing an error that occurs when converting an RPC
+    /// transaction.
+    type RpcTransactionConversionError: Debug + std::error::Error;
```
Comment (agostbiro) [inline crates/edr_evm/src/chain_spec.rs:71]: `std::error::Error` already requires `Debug`, so I think the explicit `Debug` bound could be removed.

**Review by agostbiro — APPROVED** [review]
Thanks for this PR @Wodann, it's awesome work!

Re reviewing: it's too much to go over line by line, but after @Wodann's walkthrough I did a scan + played around with it a bit locally and this + knowing @Wodann's work gives me high confidence to approve.

File: crates/edr_opt/Cargo.toml (hunk 2 above)
Comment (Wodann) [reply crates/edr_opt/Cargo.toml]: Addressed by https://github.com/NomicFoundation/edr/pull/534/commits/597c4b860e38635ff44cb9904fec13dff2ae7ad5

File: crates/edr_opt/tests/full_block.rs (hunk 1 above)
Comment (Wodann) [reply crates/edr_opt/tests/full_block.rs]: Addressed by https://github.com/NomicFoundation/edr/pull/534/commits/cbdb02b85bf632a5d49e379fa5407a8dd39e55e3

File: crates/edr_evm/src/chain_spec.rs (hunk 3 above)
Comment (Wodann) [reply crates/edr_evm/src/chain_spec.rs]: Addressed by https://github.com/NomicFoundation/edr/pull/534/commits/cbdb02b85bf632a5d49e379fa5407a8dd39e55e3

**Pull request merged**